File: java/solitaire/klondike/src/main/java/org/silnith/game/solitaire/move/filter/StockPileRecycleMustBeFollowedByAdvance.java
package org.silnith.game.solitaire.move.filter;

import java.util.Iterator;
import java.util.List;

import org.silnith.game.GameState;
import org.silnith.game.solitaire.Board;
import org.silnith.game.solitaire.move.SolitaireMove;
import org.silnith.game.solitaire.move.StockPileAdvanceMove;
import org.silnith.game.solitaire.move.StockPileRecycleMove;

public class StockPileRecycleMustBeFollowedByAdvance implements SolitaireMoveFilter {

    @Override
    public Object getStatisticsKey() {
        return "Stock Pile Recycle Must Be Followed By Advance";
    }

	@Override
	public boolean shouldFilter(final List<GameState<SolitaireMove, Board>> state) {
        final Iterator<GameState<SolitaireMove, Board>> iterator = state.iterator();
        assert iterator.hasNext();
        final GameState<SolitaireMove, Board> currentGameState = iterator.next();
        final SolitaireMove currentMove = currentGameState.getMove();
        
        if (!iterator.hasNext()) {
            return false;
        }
        
        final GameState<SolitaireMove, Board> previousGameState = iterator.next();
        final SolitaireMove previousMove = previousGameState.getMove();
		
		if (previousMove instanceof StockPileRecycleMove) {
			if (currentMove instanceof StockPileAdvanceMove) {
				// This is acceptable, no need to filter.
				return false;
			} else {
				// Why do something not involving the stock pile after recycling it?
				return true;
			}
		} else {
			// This filter doesn't apply.
			return false;
		}
	}

}

File: csharp/Solitaire/Klondike/Move/Filter/StockPileRecycleMustBeFollowedByAdvance.cs
using System.Collections.Generic;

namespace Silnith.Game.Klondike.Move.Filter
{
    public class StockPileRecycleMustBeFollowedByAdvance : ISolitaireMoveFilter
    {
        /// <inheritdoc/>
        public object StatisticsKey
        {
            get;
        } = "Stock Pile Recycle Must Be Followed By Advance";

        /// <inheritdoc/>
        public bool ShouldFilter(IReadOnlyList<GameState<ISolitaireMove, Board>> gameStateHistory)
        {
            IEnumerator<GameState<ISolitaireMove, Board>> enumerator = gameStateHistory.GetEnumerator();
            _ = enumerator.MoveNext();
            GameState<ISolitaireMove, Board> currentGameState = enumerator.Current;
            ISolitaireMove currentMove = currentGameState.Move;

            if (!enumerator.MoveNext())
            {
                /*
                 * This can only happen at the very beginning of the game.
                 * In that case, this filter is not helpful, so just let everything pass.
                 */
                return false;
            }

            GameState<ISolitaireMove, Board> previousGameState = enumerator.Current;
            ISolitaireMove previousMove = previousGameState.Move;

            if (previousMove.IsStockPileRecycle)
            {
                if (currentMove.IsStockPileModification)
                {
                    /*
                     * This is accepetable, no need to filter.
                     */
                    return false;
                }
                else
                {
                    /*
                     * Why do something not involving the stock pile after recycling it?
                     */
                    return true;
                }
            }
            else
            {
                /*
                 * This filter doesn't apply.
                 */
                return false;
            }
        }
    }
}
